### DISCUSSION TITLE
Different vizantis for different robots

### DISCUSSION THREAD

**Discussion by gxomalis** [MoffKalast/vizanti#69] [General]
Hello there! @MoffKalast 
So let's say I have robot A and robot B. What should I do if i want to have 2 separate vizantis for each robot?
Like I wanna launch a vizanti for robot A, and another vizanti for robot B, so I can have different widgets for each robot, different configuration etc. 
Is this possible? What conflicts could occur from having 2 vizantis running in one system?

**Comment by MoffKalast** [comment]
Hey, well what I usually do for that sort of setup is have the Vizanti backend running on the robots themselves, so the browser connects to different IPs and will save widgets for them separately (also handy when going somewhere and switching them to AP mode). Of course with rosbridge that does add a fair bit of overhead on ROS 2 which may not be best on low power ARM machines.

I suppose it somewhat depends if the robots are on the same ROS network (i.e. same roscore/domain_id). If yes, then you could run one Vizanti backend locally and open two web app instances in the browser and load up different widget configs (global settings -> import/export). There's generally nothing wrong with running multiple at the same time, but it will put more load on rosbridge/rws.

There's generally three parts you have to consider (usually they're rolled together into one centralized stack):
- web server (what the browser accesses for the static files, flask server on port 5000 by default)
- websocket bridge (what the web app connects to to get ROS data, either rosbridge or rws on port 5001)
- service nodes (for processing custom system functionally, they need to just be on the same ROS network)

Now theoretically you could also put just the flask server on the robot, and run the websocket bridge locally which would have zero impact on robot performance, but the IP is [currently a bit hardcoded](https://github.com/MoffKalast/vizanti/blob/1f3ddf91d94398cee83ffb65eb3d80af7cdc65ee/vizanti_server/public/js/modules/rosbridge.js#L109) to the address that's serving the web app, so you'd have to change that. As for the service nodes, they could run anywhere but that will affect what they do (e.g. you can only kill nodes on the system they're running).

Another option that might work is by using a different [baseurl config](https://github.com/MoffKalast/vizanti/blob/1f3ddf91d94398cee83ffb65eb3d80af7cdc65ee/vizanti_server/launch/vizanti_server.launch.py#L9) to launch two different namespaced instances of the flask server locally,  which might be enough to convince the browser that these are different URLs with separate localstorage. I haven't really ever tried that out though, so I'm not sure how it behaves.

**Reply by gxomalis** [reply]
That's great info thanks!
What if I want to run a vizanti on port 5000 and the other one on let's say 6000?
I want the have two or more vizantis running on different ports, with the same rosbridge, and then to customize each vizanti how ever I want and get them with an <iframe> in my web app. 
What are the steps to follow so I can achieve that?

**Reply by MoffKalast** [reply]
I've given it a short test and it looks like it should be alright if you just duplicate the flask node. So for noetic that would be something as follows:

	<node name="vizanti_flask_node" pkg="vizanti" type="server.py" output="screen">
		<param name="host" value="0.0.0.0"/>
		<param name="port" value="5000"/>
		<param name="port_rosbridge" value="$(arg port_rosbridge)"/>
		<param name="base_url" value="$(arg base_url)"/>
		<param name="flask_debug" value="$(arg flask_debug)"/>
	</node>
	
	<node name="vizanti_flask_node2" pkg="vizanti" type="server.py" output="screen">
		<param name="host" value="0.0.0.0"/>
		<param name="port" value="5050"/>
		<param name="port_rosbridge" value="$(arg port_rosbridge)"/>
		<param name="base_url" value="$(arg base_url)"/>
		<param name="flask_debug" value="$(arg flask_debug)"/>
	</node>

And on humble:

    flask_node = launch_ros.actions.Node(
        name='vizanti_flask_node',
        package='vizanti_server',
        executable='server.py',
        output='screen',
        parameters=[
            {'host': '0.0.0.0'},
            {'port': 5000},
            {'port_rosbridge': port_rosbridge},
            {'flask_debug': flask_debug},
            {'base_url': base_url},
            {'compression': "none"}
        ]
    )

    flask_node2 = launch_ros.actions.Node(
        name='vizanti_flask_node2',
        package='vizanti_server',
        executable='server.py',
        output='screen',
        parameters=[
            {'host': '0.0.0.0'},
            {'port': 5050},
            {'port_rosbridge': port_rosbridge},
            {'flask_debug': flask_debug},
            {'base_url': base_url},
            {'compression': "none"}
        ]
    )

The rest of the launch files can remain as-is. The browser does treat different ports as different sites, so this should save separate configs.

**Reply by gxomalis** [reply]
Great that worked!
Thanks for the help, is it okay to leave the discussion open in case I need further assistance in the future?

**Reply by MoffKalast** [reply]
Sure thing :)
